parallel-twitter-trending-hashtags/twitter_wallaroo_app.py

```diff
@@ -0,0 +1,173 @@
+    """
+    receive a dictionary of tags and counts,
+    add to our list of top tags and shrink to
+    only ten, if our new top ten is different than
+    our old, output the new list
+    """
+    top_ten_if_changed = top_tags.update(tags_and_counts)
+    return (top_ten_if_changed, top_ten_if_changed != None)
+
+
+class HashtagCounts(object):
+    def __init__(self):
+        self.__top_tags = {}
+        self.__size = TRENDING_OVER
+        self.__window = []
+        for x in range(0, self.__size):
+            self.__window.append((0, {}))
+
+    def increment(self, hashtag):
+        """
+        Increment the count for `hashtag`
+        """
+        mse = self.__minutes_since_epoch()
+        window_gap = int(mse % TRENDING_OVER)
+
+        # have we rolled over?
+        if self.__window[window_gap][0] < mse:
+            self.__rollover(hashtag, window_gap, mse)
+        else:
+            self.__increment(hashtag, window_gap)
+
+        # did our top ten change?
+        new_top_tags = self.__calculate_top_tags()
+
+        if new_top_tags != self.__top_tags:
+            self.__top_tags = new_top_tags
+            return self.__top_tags.copy()
+        else:
+            return None
+
+    def __minutes_since_epoch(self):
+        return int(time.time() / 60)
+
+    def __increment(self, hashtag, gap):
+        current_count = self.__window[gap][1].get(hashtag, 0)
+        self.__window[gap][1][hashtag] = current_count + 1
+
+    def __rollover(self, hashtag, gap, mse):
+        self.__window[gap] = (mse, {hashtag: 1})
+
+    def __calculate_top_tags(self):
+        tags = {}
+        for x in range(0, self.__size):
+            subset = self.__window[x][1]
+            for (key, value) in subset.items():
+                tags[key] = tags.get(key, 0) + value
+
+        return dict(sorted(tags.items(), key=lambda (k, v): v,
+                           reverse=True)[:TOP_X])
+
+
+class TopTags(object):
+    def __init__(self):
+        self.__top_tags = {}
+
+    def update(self, hashtags_and_counts):
+        self.__top_tags.update(hashtags_and_counts)
+        new_top_tags = self.__calculate_top_tags()
+
+        top_tags_changed = False
+
+        if new_top_tags != self.__top_tags:
+            top_tags_changed = True
+
+        # trim to TOP_X
+        self.__top_tags = new_top_tags
+
+        if top_tags_changed:
+            return self.__top_tags.copy()
+        else:
+            return None
+
+    def __calculate_top_tags(self):
+        return dict(sorted(self.__top_tags.items(), key=lambda (k, v): v,
+                           reverse=True)[:TOP_X])
+
+
+@wallaroo.decoder(header_length=4, length_fmt=">I")
+def decoder(bs):
+    return bs.decode("utf-8")
+
+
+@wallaroo.partition
+def extract_hashtag_key(data):
+    if data[0] >= "a" and data[0] <= "z":
+        return data[0]
+    else:
+        return "!"
+
+
+@wallaroo.encoder
+def encoder(data):
+    top_first = sorted(data.items(), key=lambda (k, v): v, reverse=True)
```

Comment:
I think that the sorting should be in its own computation, rather than in the encoder.

Reply:
i'm not changing it at this point, i need to finish the post. i don't have time to do any changes to logic etc and make sure they work.